flagext/map.go

```diff
@@ -0,0 +1,108 @@
+package flagext
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"gopkg.in/yaml.v3"
+)
+
+// LimitsMap is a flag.Value implementation that looks like a generic map, holding float64s, ints, or strings as values.
+type LimitsMap[T float64 | int | string] struct {
+	data      map[string]T
+	validator func(k string, v T) error
+}
+
+func NewLimitsMap[T float64 | int | string](validator func(k string, v T) error) LimitsMap[T] {
+	return NewLimitsMapWithData(make(map[string]T), validator)
+}
+
+func NewLimitsMapWithData[T float64 | int | string](data map[string]T, validator func(k string, v T) error) LimitsMap[T] {
+	return LimitsMap[T]{
+		data:      data,
+		validator: validator,
+	}
+}
+
+// IsInitialized returns true if the map is initialized.
+func (m LimitsMap[T]) IsInitialized() bool {
+	return m.data != nil
+}
+
+// String implements flag.Value
+func (m LimitsMap[T]) String() string {
+	out, err := json.Marshal(m.data)
+	if err != nil {
+		return fmt.Sprintf("failed to marshal: %v", err)
+	}
+	return string(out)
+}
+
+// Set implements flag.Value
+func (m LimitsMap[T]) Set(s string) error {
+	newMap := make(map[string]T)
+	if err := json.Unmarshal([]byte(s), &newMap); err != nil {
+		return err
+	}
+	return m.updateMap(newMap)
+}
+
+func (m LimitsMap[T]) Read() map[string]T {
+	return m.data
+}
+
+// Clone returns a copy of the LimitsMap.
+func (m LimitsMap[T]) Clone() LimitsMap[T] {
+	newMap := make(map[string]T, len(m.data))
+	for k, v := range m.data {
+		newMap[k] = v
+	}
+	return LimitsMap[T]{data: newMap, validator: m.validator}
+}
+
+// UnmarshalYAML implements yaml.Unmarshaler.
+func (m LimitsMap[T]) UnmarshalYAML(value *yaml.Node) error {
+	newMap := make(map[string]T)
+	if err := value.Decode(newMap); err != nil {
+		return err
+	}
+	return m.updateMap(newMap)
+}
+
+func (m LimitsMap[T]) updateMap(newMap map[string]T) error {
```

Comment:
Does this need to have a pointer receiver?

Reply:
Not strictly (though I see an argument for doing it to prevent footguns in future changes)

It doesn't change a field in the LimitsMap struct directly; it only changes the contents of the map pointed to by `m.data`. You can copy the outer struct freely and the copies will still point to the same underlying map.